### PR TITLE
[Diagnostics] Improve diagnostics for type parameter requirements

### PR DESCRIPTION
**Pull request by xedin**
Detect and diagnose failures in constraints representing
type parameter requirements generated by `openGeneric`.

Resolves: rdar://problem/35890334

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -1668,6 +1668,73 @@ bool FailureDiagnosis::diagnoseConversionToBool(Expr *expr, Type exprType) {
...
+    return false;
+
+  ownerType = cs.simplifyType(ownerType)->getWithoutSpecifierType();
+  if (ownerType->hasTypeVariable() || ownerType->hasUnresolvedType())
+    return false;
+
+  auto lhs = cs.simplifyType(constraint->getFirstType());
+  auto rhs = cs.simplifyType(constraint->getSecondType());
+
+  switch (constraint->getKind()) {
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: Do we not have a similar switch elsewhere?

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
Weird, `smoke test` triggered Linux but not macOS for some reason...

**Pull request merged**